Fix inconsistent behavior of net sales in analytics

louwie17 opened the pull request:
### Changes proposed in this Pull Request:

This is the re-introduction of https://github.com/woocommerce/woocommerce/pull/53019 with two main changes:
- Removal of the `get_total_shipping_tax_refunded` function introduced in the `WC_Order_Data_Store_Interface` and instead added the function to the abstract-wc-order-data-store-cpt.php class.
- Moved DB key update logic before `dbDelta( self::get_schema() )` is run, to prevent this error from happening:
```
WordPress database error Multiple primary key defined for query ALTER TABLE wp_wc_order_product_lookup ADD PRIMARY KEY  (`order_item_id`,`order_id`) made by require_once('wp-load.php'), require_once('wp-config.php'), require_once('wp-settings.php'), do_action('init'), WP_Hook->do_action, WP_Hook->apply_filters, WC_Install::check_version, WC_Install::install, WC_Install::install_core, WC_Install::create_tables, dbDelta
```

Closes https://github.com/woocommerce/woocommerce/issues/32837 https://github.com/woocommerce/woocommerce/issues/48761 https://github.com/woocommerce/woocommerce/issues/48775.

(For Bug Fixes) Bug introduced in PR # .

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

**Note: The refund order stuff should of already been tested as part of the [original PR](https://github.com/woocommerce/woocommerce/pull/53019) that was suppose to be release with 9.9, but got reverted because of the backwards compatibility issue with the interface changes**.

1. Before loading this branch make sure you have the latest WooCommerce installed and activated.
2. Double check with `phpMyAdmin` if the `wp_wc_order_product_lookup` has a single primary key ( `order_item_id` ).
3. Now load this branch ( or install it, you can generate a zip using: `pnpm --filter=@woocommerce/plugin-woocommerce build:zip` )
4. Check if the `wp_wc_order_product_lookup` has two primary keys now and check if there are **no** error logs.
5. Install & activate this plugin [custom-order-store.zip](https://github.com/user-attachments/files/20694955/custom-order-store.zip) which implements the `WC_Order_Data_Store_Interface` interface, and make sure no fatals happen when navigating around.
6. Now create several completed orders with a shipping amount and change the dates to something in the past
7. Refund and partial refund a couple of the orders ( partial refund can be done [this way](https://github.com/woocommerce/woocommerce/assets/1314156/0c5c5781-2a7d-4450-bedb-cf75d03de335) )
8. Go to **WooCommerce > Status > Scheduled Actions** and make sure there are only two ( or less ) pending actions.
9. Go to **Analytics > Revenue** and take note of the Net Sales
<img width="1307" alt="Screenshot 2025-06-11 at 14 08 30" src="https://github.com/user-attachments/assets/4c5c9bb2-f9ab-4541-ae62-a762a3a46e7c" />
10. Look at refunds, this should be the total refunded amount including the tax amount refunded ( this was different for full/partial refunds in the past )
11. If you created an order and selected a full refund the same day, then the Net sales and taxes should render zero for that day ( previously they didn't ).
12. Go to **Analytics > Orders** and **Analytics > Products** the Net sales amount should be the same as the Revenue one ( when comparing to products, you may need to subtract any fees from an order, the smooth generator adds this sometimes ).
13. Now change the date ranges around and compare the above two, they should stay the same.

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

ianlin committed:
Fix incorrect net sales in analytics product reports for partial and full refunds (#53019)

* Fix inconsistent behavior for Net Sales in Analytics for single line refund (#48768)

* Fix behavior for single line refund

* Add comments

* Add hook

* Change version trigger

* Add changelog

* Fix net_revenue

* Revert gross_revenue changes

* Update version to include the fix

* Update target release version to 9.4

---------

Co-authored-by: Ian Yu-Hsun Lin <ianlin@automattic.com>

* Fix inconsistent behaviour for Net Sales in Analytics for full refund (#51262)

* Update primary key to composite (order_item_id, order_id) in the wc_order_product_lookup table

* Add full refunded order record to wc_order_product_lookup table

* Delete order items in order product lookup when the order is full refund

The previous commit will add the order item with negative net revenue
when the order is fully refunded. But the items sold in the products
analytics page is incorrect, because there was a DB record of an order item with
positive net revenue, and the product analytics query simply sums up all
the positive net revenue.

* Revert "Update primary key to composite (order_item_id, order_id) in the wc_order_product_lookup table"

This reverts commit 785a66521797ab14e84950b27a53f6ef941aacd7.

* Delete order item with negative sales recorded earlier due to partial refund

* Extract the logic of delete full refunded orders to a separate method

* Add changefile(s) from automation for the following project(s): woocommerce

* Revert "Extract the logic of delete full refunded orders to a separate method"

This reverts commit 47325f69d6f549d3c591270a877647594c403d9a.

* Revert "Delete order item with negative sales recorded earlier due to partial refund"

This reverts commit 18751d619364c2a5ad8ed1193da30a858eb04ea1.

* Revert "Delete order items in order product lookup when the order is full refund"

This reverts commit 1af8158e351b5f2bbe09796e2c1e367bf719eb04.

* Revert "Revert "Update primary key to composite (order_item_id, order_id) in the wc_order_product_lookup table""

This reverts commit b682c393bfccdb426fabf49e6704a5075d60f54b.

* Remove the partially refunded order item if exists

* Update analytics order stats test

Since we have added refunded order line items in wc_order_product_lookup
table, the unique product count should also relect the amount of
refunded products.

* Ignore refunded products in extended info for analytic orders report

---------

Co-authored-by: github-actions <github-actions@github.com>

* Add changefile(s) from automation for the following project(s): woocommerce

* Hook refunded actions and add refund type to orders meta

* Update primary key to composite (order_item_id, order_id, product_id) in the wc_order_product_lookup table

* Get the parent order items when it's a full refund, and exclude the
already partially refunded items.

* Update order stats report tests by adding a line item in the refund

* Add tests for full refunds and shipping fee refunds in product reports

* Update primary key to composite (order_item_id, order_id, product_id) in the wc_order_product_lookup table

* Update migration SQL for refunded orders not presented in product loolup
table

* Check for empty order items when full refund is triggered by changing
order status to "Refunded"

* Change the primary composite key back to (order_item_id, order_id) and
do not store 0 for order_item_id when the full refund does not have
order line items

Because in this case although the order_item_id will be repeated but order_id
will be the refund order ID, the combination of them will still be
unique.

* Handle partially refunded products

If a single line item was refunded 60% then fully refunded after, we need store the difference in the product lookup table.
E.g. A product costs $100, it was previously partially refunded $60, then fully refunded $40.
So it will be -abs( 100 + (-60) ) = -40.;

* Handle the case when there are multiple partially refunded products in
an order

* Handle variation products for full refund without line items

* Fix PHPCS

* Use %i as the placeholder for the table name in the prepare query

The prepare placeholder "%i" was released in WP 6.3, and the current minimum WP version for WC is 6.5.

* Handle the old refund orders don't have _refund_type metadata in
migration script

* Remove condition checking for refunded items in product lookup table in
migration script

* Update migration scripts version to WC 9.7

* Add a test for when all line items in an order are refunded each by 100%

* Simplify the code for the case of full refunds without line items

* Calculate shipping amount when the order status chanced to REFUNDED

i.e. A full refund without order line items.

* Handle the case when refund an order item then do a full refund

The negative shipping fee should be distributed based on the number of remaining
items left to refund, rather than the number of total order items.

* Add a fallback to set to 0 for the partial_refund_product_revenue lookup

* Calculate shipping tax amount when the order status changed to REFUNDED

i.e. A full refund without order line items.

* Fix wpdb prepare for IN clause for OrdersTableDataStore::get_total_tax_refunded

* Calculate cart tax amount when the order status changed to REFUNDED

i.e. A full refund without order line items.

* Fix PHPCS

* Set coupon_amount to 0 when the order status changed to REFUNDED

i.e. A full refund without order line items.

* Create a function get_total_shipping_tax_refunded to ensure the cache is
different from get_total_tax_refunded

* Add a test for the function get_total_shipping_tax_refunded()

* Check the passed argument instead of the one init by it

* Move the lines within the if condition as it's only used in it

* Makes the method get_item_cart_tax_amount() return float consistently

* Add comments in OrderTraits for the scenario of partial refunds in an
order

* Refactor: get cart tax amount when it's set instead of adding all up

Since an order line item has only one tax rate.

* Fallback to return 0 when the method get_total_shipping_tax_refunded
doesn't exist in the datastore

* Revert "Refactor: get cart tax amount when it's set instead of adding all up"

This reverts commit 40e9f294d8e7cb50d0d8ba6e767ba7b7ba12116a.

* Add a test for full refund compound tax in analytics sync orders

* Rename DB migration functions so they are targeting WC 9.9

* Pass a variable to end() function instead of a function call

* Use $wpdb->prefix in the query for unit tests

* Add changefile(s) from automation for the following project(s): woocommerce

* Update @since tags to 9.9.0

* Merge the changelogs to clarify the fix only focuses on product reports

* Add changefile(s) from automation for the following project(s): woocommerce

* Remove the auto-generated changelog

---------

Co-authored-by: Fernando Marichal <fernando.marichal@automattic.com>
Co-authored-by: github-actions <github-actions@github.com>

louwie17 committed:
Update version of update functions

louwie17 committed:
Update new primary key addition

louwie17 committed:
Only run primary key update if table already exists

louwie17 committed:
Move get_total_shipping_tax_refunded to abstract order data store class

louwie17 committed:
Remove function from interface

louwie17 committed:
Fix update logic

louwie17 commented on plugins/woocommerce/includes/class-wc-install.php:
```diff
@@ -1574,6 +1576,12 @@ public static function create_tables() {
+			if ( 2 > $wpdb->get_var( "SELECT COUNT(*) FROM INFORMATION_SCHEMA.STATISTICS WHERE TABLE_SCHEMA = DATABASE() AND TABLE_NAME = '{$wpdb->prefix}wc_order_product_lookup' AND INDEX_NAME = 'PRIMARY'" ) ) {
+				$result = $wpdb->query( "ALTER TABLE {$wpdb->prefix}wc_order_product_lookup DROP PRIMARY KEY, ADD PRIMARY KEY (order_item_id, order_id)" );
+			}
+		}
```
Moved here so it can run before `$db_delta_result = dbDelta( self::get_schema() );`

chihsuan replied:
If we moved this here, why do we still need `wc_update_1000_update_primary_key_to_composite_in_order_product_lookup_table` in db updates?

Additionally, it would be helpful to include some comments to explain why we add it here. 🙂

louwie17 replied:
Thanks, makes sense! I removed it.

Nikschavan replied:
Will adding it here will drop the index and re-create it for each WC update? On large stores with many product items sold the size of this table could be large and this can be an expensive operation

Nikschavan replied:
Never mind, I just realised we are indeed checking the number of columns for the primary key

chihsuan commented on plugins/woocommerce/includes/wc-update-functions.php:
```diff
@@ -2993,6 +2993,58 @@ function wc_update_961_migrate_default_email_base_color() {
+			 * Trigger an action to schedule the data import for old refunded order items.
+			 *
+			 * @param int $order_id The ID of the order to be synced.
+			 * @since 9.9.0
```
```suggestion
			 * @since 10.0.0
```

chihsuan commented on plugins/woocommerce/includes/class-wc-order.php:
```diff
@@ -2207,6 +2207,31 @@ public function get_total_tax_refunded() {
+	/**
+	 * Get the total shipping tax refunded.
+	 *
+	 * @since  9.9.0
```
```suggestion
	 * @since  10.0.0
```

chihsuan commented on plugins/woocommerce/includes/data-stores/abstract-wc-order-data-store-cpt.php:
```diff
@@ -805,4 +805,30 @@ protected function update_order_meta_from_object( $order ) {
+	 *
+	 * @param  WC_Order $order Order object.
+	 *
+	 * @since 9.9.0
```
```suggestion
	 * @since 10.0.0
```

chihsuan commented on plugins/woocommerce/includes/data-stores/abstract-wc-order-data-store-cpt.php:
```diff
@@ -805,4 +805,30 @@ protected function update_order_meta_from_object( $order ) {
+				AND order_itemmeta.meta_key = 'shipping_tax_amount'",
+				$order->get_id()
+			)
+		) ?? 0;
```
Since we're using inner joins here, I think there is an opportunity to optimize the query for better performance.

Need to double-check, but here is the revised SQL query:

```sql
$total = $wpdb->get_var( 
    $wpdb->prepare( 
        "SELECT COALESCE(SUM(order_itemmeta.meta_value), 0) 
         FROM {$wpdb->prefix}woocommerce_order_items AS order_items
         INNER JOIN {$wpdb->prefix}woocommerce_order_itemmeta AS order_itemmeta 
            ON order_items.order_item_id = order_itemmeta.order_item_id
         INNER JOIN {$wpdb->posts} AS posts 
            ON order_items.order_id = posts.ID
         WHERE posts.post_type = 'shop_order_refund' 
           AND posts.post_parent = %d 
           AND order_items.order_item_type = 'tax'
           AND order_itemmeta.meta_key = 'shipping_tax_amount'
           AND order_itemmeta.meta_value IS NOT NULL", 
        $order->get_id() 
    ) 
);
```

 **Improvements Made**

1. Added COALESCE(SUM(...), 0) - handles NULL results in SQL
2. Reordered JOINs - start with order_items table first
3. Added IS NOT NULL - excludes NULL meta values from SUM
4. Moved conditions to WHERE - better performance than JOIN conditions

louwie17 replied:
Did you get a chance to test this? I just asked Cursor and it seemed to think that the current query implementation is more performant then this one, for the below reasons:

Current query is likely more performant because:
- Better JOIN order: Starting with order_itemmeta and filtering by meta_key = 'shipping_tax_amount' early reduces the dataset significantly before joining with other tables.
- Fewer conditions in WHERE: current has fewer WHERE conditions, which can be more efficient.
- Combined JOIN conditions: The combined JOIN conditions in current query can sometimes be optimized better by the query planner.

chihsuan replied:
I just ran some tests, and both queries seem to have similar performance. This may be because MySQL’s optimizer is able to reorder joins. The current query works for me.

<img width="873" height="243" alt="Screenshot 2025-08-20 at 11 35 54" src="https://github.com/user-attachments/assets/96efe64c-256b-4875-bb77-76405f700629" />
<img width="868" height="234" alt="Screenshot 2025-08-20 at 11 35 30" src="https://github.com/user-attachments/assets/728f160b-807b-4987-95c3-e345a559baeb" />

chihsuan commented on plugins/woocommerce/src/Admin/API/Reports/Orders/DataStore.php:
```diff
@@ -502,8 +502,9 @@ protected function get_products_by_order_ids( $order_ids ) {
 				)
-			WHERE
-				order_id IN ({$included_order_ids})
+			WHERE 1 = 1
```
Could we remove `1 = 1`? It seems to be unnecessary.

chihsuan commented on plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php:
```diff
@@ -1048,6 +1049,36 @@ public function get_total_tax_refunded( $order ) {
+	 *
+	 * @param WC_Order $order Order object.
+	 *
+	 * @since 9.9.0
```
```suggestion
	 * @since 10.0.0
```

chihsuan commented on plugins/woocommerce/src/Admin/Overrides/OrderTraits.php:
```diff
@@ -16,20 +16,30 @@ trait OrderTraits {
+		// Use the passed shipping_amount if provided, otherwise get the total shipping amount in the order.
+		// This is useful when calculating refunds for partial shipping in an order.
+		// For example, if $10 shipping is refunded from an order with $30 shipping, the remaining $20 should be distributed to the remaining items.
+		$total_shipping_amount = $shipping_amount ? $shipping_amount : (float) $this->get_shipping_total();
```
If $shipping_amount is 0, the method mistakenly uses the order’s shipping total. Should we use a null check instead, or check if it's numeric to avoid potential errors?

```php
$total_shipping_amount = null !== $shipping_amount
    ? $shipping_amount
    : (float) $this->get_shipping_total();
```

chihsuan commented on plugins/woocommerce/src/Admin/Overrides/OrderTraits.php:
```diff
@@ -42,29 +52,43 @@ public function get_item_shipping_amount( $item ) {
-					$tax_item_id                = $tax_item->get_rate_id();
-					$tax_item_total             = isset( $tax_data['total'][ $tax_item_id ] ) ? (float) $tax_data['total'][ $tax_item_id ] : 0;
-					$total_shipping_tax_amount += $tax_item_total;
+		// Use the passed shipping_tax_amount if provided, otherwise initiliase it to 0 and calculate the total shipping tax amount in the order.
```
typo

```suggestion
		// Use the passed shipping_tax_amount if provided, otherwise initialize it to 0 and calculate the total shipping tax amount in the order.
```

chihsuan commented on plugins/woocommerce/src/Admin/Overrides/OrderTraits.php:
```diff
@@ -16,20 +16,30 @@ trait OrderTraits {
+		// Use the passed order_items_count if provided, otherwise get the total number of items in the order.
+		// This is useful when calculating refunds for partial items in an order.
+		// For example, if 2 items are refunded from an order with 4 items. The remaining 2 items should have the shipping fee of the refunded items distributed to them.
+		$order_items = $order_items_count ? $order_items_count : $this->get_item_count();
```
If $order_items_count is 0, the method mistakenly uses the get_item_count. Should we use a null check instead, or check if it's numeric to avoid potential errors?

chihsuan replied:
@louwie17 👋 What are your thoughts on this? Same as the comment below.

louwie17 replied:
Ooops yeah, I thought I addressed them all, but missed these two, thanks :)

chihsuan commented on plugins/woocommerce/src/Admin/Overrides/OrderTraits.php:
```diff
@@ -42,29 +52,43 @@ public function get_item_shipping_amount( $item ) {
+		// Use the passed order_items_count if provided, otherwise get the total number of items in the order.
+		// This is useful when calculating refunds for partial items in an order.
+		// For example, if 2 items are refunded from an order with 4 items. The remaining 2 items should have the shipping tax of the refunded items distributed to them.
+		$order_items = $order_items_count ? $order_items_count : $this->get_item_count();
```
Same here.

chihsuan commented on plugins/woocommerce/src/Admin/API/Reports/Products/DataStore.php:
```diff
@@ -433,30 +467,126 @@ public static function sync_order_products( $order_id ) {
+		$parent_order = null;
+
+		// When changing the order status to "Refunded", the refund order's type will be full refund, and the order items will be empty.
+		// We need to get the parent order items, and exclude the items that is already being patially refunded.
```
typo

```suggestion
		// We need to get the parent order items, and exclude the items that are already being partially refunded.
```

chihsuan reviewed (commented):
Thank you for addressing this issue! 

I tested the branch and confirmed:

1. No errors in debug.log when updating from WC 9.9 and with the `custom-order-store` plugin installed.
2. Manually created orders and refunds are working as expected. The numbers in Analytics > Products are the same as in Analytics > Revenue.

However, I found that when I generated order vis `wc-smooth-generator` (without refound), the numbers in Analytics > Products are not the same as in Analytics > Revenue, or Orders. Do we know if this is expected?

https://github.com/user-attachments/assets/54b7e51c-3af5-48c8-bbc5-417ff270c367

I've also reviewed the code and left some comments.

chihsuan commented:
I'll be testing it more tomorrow.

chihsuan committed:
Update plugins/woocommerce/src/Admin/API/Reports/Products/DataStore.php

chihsuan committed:
Update plugins/woocommerce/src/Admin/Overrides/OrderTraits.php

chihsuan commented:
> However, I found that when I generated order vis wc-smooth-generator (without refound), the numbers in Analytics > Products are not the same as in Analytics > Revenue, or Orders. Do we know if this is expected?

I've confirmed that this issue doesn't occur when using WooCommerce 9.9. The new order is somehow not reflected in Analytics.


https://github.com/user-attachments/assets/76304c6a-fc5d-4e35-9e1a-eda1d8ba999a

chihsuan commented on plugins/woocommerce/includes/wc-update-functions.php:
```diff
@@ -2993,6 +2993,58 @@ function wc_update_961_migrate_default_email_base_color() {
+function wc_update_1000_add_old_refunded_order_items_to_product_lookup_table() {
+	global $wpdb;
+
+	// Get every order ID where:
```
Would it be worth testing this change in a large store to see if any performance issues arise? This could potentially affect query execution time and resource usage, as we will be retrieving all old refunded order items from the product lookup table.

louwie17 replied:
Yeah it did probably be worth testing, as I am looking at this logic, it also looks like this can probably be optimized quite a bit, as we are doing:
- Getting all the refunded order ids
- Then getting each order object individually ( triggering a query in behind )
- Checking if the order has line items ( this could be checked in the original query )
- Triggering a scheduled action for each refunded order ( this may not be as bad, as the scheduler will handle this ).

louwie17 replied:
So I did make a small update to this, so we only retrieve the entire order object if the `num_items_sold` is set to 0, this more accurately reflects a fully refunded order.
I don't anticipate the actual select query to be much of a bottleneck.

Plus this update function gets triggered in a cron job, same goes for the import calls, which in turn also gets badged if there is a bunch of them.

nerrad commented:
@louwie17 and @chihsuan, it looks like this got stalled, what are the plans here?

louwie17 committed:
Address PR feedback

louwie17 committed:
Update logic to more accurately target full refund orders

louwie17 committed:
Update num_items_sold correctly on full refunds

louwie17 committed:
Fix lint issues

chihsuan commented on plugins/woocommerce/src/Admin/API/Reports/Products/DataStore.php:
```diff
@@ -437,30 +471,126 @@ public static function sync_order_products( $order_id ) {
+			$parent_order    = wc_get_order( $parent_order_id );
+			$order_items     = $parent_order->get_items();
+
+			// Get the partially refunded product and variation IDs along with their sum of product_net_revenue from the parent order.
+			$partial_refund_products = $wpdb->get_results(
+				$wpdb->prepare(
+					"
+						SELECT
+							product_lookup.product_id,
+							product_lookup.variation_id,
+							SUM( product_lookup.product_net_revenue ) AS product_net_revenue
+						FROM %i AS product_lookup
+						INNER JOIN {$wpdb->prefix}wc_order_stats AS order_stats
+							ON order_stats.order_id = product_lookup.order_id
+						WHERE 1 = 1
+							AND order_stats.parent_id = %d
+							AND product_lookup.product_net_revenue < 0
+						GROUP BY product_lookup.product_id, product_lookup.variation_id
+					",
+					$table_name,
+					$parent_order_id
+				)
+			);
```
Since WordPress 6.2, `$wpdb->prepare()` does support %i as an identifier placeholder (for table/column names).

chihsuan commented on plugins/woocommerce/src/Admin/API/Reports/Products/DataStore.php:
```diff
@@ -112,6 +112,40 @@ protected function assign_report_columns() {
+	 * @param int    $refund_id Refund ID.
+	 * @param string $type      Refund type.
+	 */
+	public static function add_refund_type_meta( $refund_id, $type ) {
+		$order = wc_get_order( $refund_id );
+		$order->update_meta_data( '_refund_type', $type );
+		$order->save_meta_data();
 	}
```
Since this function is invoked by hooks, it's generally safe to assume the order exists.

chihsuan reviewed (commented):
I've tested this PR again on a fresh JN and confirmed that the analytics net sales calculations now work correctly across Revenue, Orders, and Products reports. 👍 

Only left two inline comments in the code regarding how zero values are handled. Instead of relying on truthy checks, I suggest we explicitly compare the value to null.

The unit tests seem to be failing due to the fix for full refound logic. I think it's an expected behavior and we should update expectations to reflect the new behavior, since we intentionally change how full refunds affect `num_items_sold`.

louwie17 committed:
Update avg items and avg order value to not include refunds

louwie17 committed:
Update test with correct data

louwie17 committed:
Address null checks

louwie17 committed:
Remove log debugging statement

louwie17 committed:
Fix unit tests

louwie17 commented:
Thanks for the review @chihsuan :)

> Only left two inline comments in the code regarding how zero values are handled. Instead of relying on truthy checks, I suggest we explicitly compare the value to null.

Thanks, I missed those prior, and addressed them this time!

> The unit tests seem to be failing due to the fix for full refound logic. I think it's an expected behavior and we should update expectations to reflect the new behavior, since we intentionally change how full refunds affect `num_items_sold`.

While looking at the tests, I realized a couple other things that didn't make sense. For example in the stats the `orders` count skips refunded orders, but the `avg_order_value` and `avg_items_per_order` does include them, but both of them only divide by non refunded orders.
So I updated the query in this commit: https://github.com/woocommerce/woocommerce/pull/58744/commits/f95e14e011c06e27f008f3c5a08465cf130487f8 and updated the tests accordingly.

chihsuan reviewed (approved):
> While looking at the tests, I realized a couple other things that didn't make sense. For example in the stats the orders count skips refunded orders, but the avg_order_value and avg_items_per_order does include them, but both of them only divide by non refunded orders.
So I updated the query in this commit: https://github.com/woocommerce/woocommerce/commit/f95e14e011c06e27f008f3c5a08465cf130487f8 and updated the tests accordingly.

Looks good to me! Thanks for catching that inconsistency.

louwie17 committed:
Set correct tax_total and net_total for full refunds

louwie17 committed:
Update refunds and gross sales correctly

louwie17 committed:
Fix linting

louwie17 committed:
Remove full refund tooltip, as it is not relevant anymore

chihsuan reviewed (commented):
> * For full refunds we also set the correct `tax_total` and `net_total` ( not including the tax ): [f6da4ef](https://github.com/woocommerce/woocommerce/commit/f6da4ef9b61e4fbc0de7ee1f883737fe527554d1) this was done correctly in the partial refunds
> * Updated the refunds calculation to include the tax amount, this required a small change in the gross total as well: [4de3530](https://github.com/woocommerce/woocommerce/commit/4de35304cd3f46f439f3ded5f2826e51ebd7b20f)
> * Removed the full refund not being included tooltip [ddba17e](https://github.com/woocommerce/woocommerce/commit/ddba17e7f31cc9126e98f92efabd8612f2834a00)
> 
> This should fix all the issues in regards to full refunds I believe, and will also be relevant for Woo Analytics, since we rely on these tables.

Thank you for the explanation! @louwie17 It helps me understand the changes. :slightly_smiling_face:

I tested a full refund and noticed that the "Return" value in the report does not match the actual refunded amount. It's because shipping costs ($1.00) are not included. Should we also include shipping costs in refund calculations?

I full-refounded an order ($11.01 = $10.00 subtotal + $1.00 shipping + $0.01 tax).
<img width="952" height="474" alt="Screenshot 2025-08-22 at 11 28 28" src="https://github.com/user-attachments/assets/9fa3e304-992e-42bb-b489-6df6db61ad53" />

"Return" in the report is `$10.1` instead of `$11.01`:
<img width="1150" height="493" alt="Screenshot 2025-08-22 at 11 29 13" src="https://github.com/user-attachments/assets/f0066c15-f3dd-457a-a82a-7b370f58ad52" />

louwie17 committed:
Make sure shipping_total is set for refunds

louwie17 committed:
Add shipping total to refund calculation, and add tooltip for this

louwie17 committed:
Update tooltip to use returns instead of refunds

louwie17 commented:
> I tested a full refund and noticed that the "Return" value in the report does not match the actual refunded amount. It's because shipping costs ($1.00) are not included. Should we also include shipping costs in refund calculations?
> 
> I full-refounded an order ($11.01 = $10.00 subtotal + $1.00 shipping + $0.01 tax).

Thanks @chihsuan for the additional review, and catching the shipping costs refunds. That is a good suggestion and another item that the full refunds wasn't updating correctly.
I made three follow up changes:
- Update the import script so that the correct shipping total is set for full refunds ( this may require a trigger of DB update again to set the correct data ): https://github.com/woocommerce/woocommerce/pull/58744/commits/af2ed12a53b7e5b3a023d1bae8920004bef36ccc ( partial refunds did this correctly.
- Added `shipping_total` to the sum of the refunds query https://github.com/woocommerce/woocommerce/pull/58744/commits/5f95a8d472c0942ac9654fd477930a230530eef7
- Added a tooltip to the returns box to mention that it includes returned tax/shipping amounts ( as this was not explicit before ) https://github.com/woocommerce/woocommerce/pull/58744/commits/e87a74ff1e6c67e0ce317d382e45d23798d10cfe

chihsuan reviewed (approved):
> Update the import script so that the correct shipping total is set for full refunds ( this may require a trigger of DB update again to set the correct data ): https://github.com/woocommerce/woocommerce/commit/af2ed12a53b7e5b3a023d1bae8920004bef36ccc ( partial refunds did this correctly.
Added shipping_total to the sum of the refunds query https://github.com/woocommerce/woocommerce/commit/5f95a8d472c0942ac9654fd477930a230530eef7
Added a tooltip to the returns box to mention that it includes returned tax/shipping amounts ( as this was not explicit before ) https://github.com/woocommerce/woocommerce/commit/e87a74ff1e6c67e0ce317d382e45d23798d10cfe

Thanks for addressing this! @louwie17 Tested this locally and it looks good to me. 🎉
